pkg/cmd/server/origin/handlers.go

```diff
@@ -291,3 +297,112 @@ func assetServerRedirect(handler http.Handler, assetPublicURL string) http.Handl
 		handler.ServeHTTP(w, req)
 	})
 }
+
+func (c *MasterConfig) impersonationFilter(handler http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		requestedSubject := req.Header.Get("Impersonate-User")
+		if len(requestedSubject) == 0 {
+			handler.ServeHTTP(w, req)
+			return
+		}
+
+		resource, namespace, name, err := parseRequestedSubject(requestedSubject)
+		if err != nil {
+			forbidden(err.Error(), nil, w, req)
+			return
+		}
+
+		ctx, exists := c.RequestContextMapper.Get(req)
+		if !exists {
+			forbidden("context not found", nil, w, req)
+			return
+		}
+
+		actingAsAttributes := &authorizer.DefaultAuthorizationAttributes{
+			Verb:         "impersonate",
+			APIGroup:     resource.Group,
+			Resource:     resource.Resource,
+			ResourceName: name,
+		}
+		authCheckCtx := kapi.WithNamespace(ctx, namespace)
+
+		allowed, reason, err := c.Authorizer.Authorize(authCheckCtx, actingAsAttributes)
+		if err != nil {
+			forbidden(err.Error(), actingAsAttributes, w, req)
+			return
+		}
+		if !allowed {
+			forbidden(reason, actingAsAttributes, w, req)
+			return
+		}
+
+		switch resource {
+		case kapi.Resource(authorizationapi.ServiceAccountResource):
+			newUser := &user.DefaultInfo{
+				Name:   serviceaccount.MakeUsername(namespace, name),
+				Groups: serviceaccount.MakeGroupNames(namespace, name),
+			}
+			newUser.Groups = append(newUser.Groups, "system:authenticated")
+			c.RequestContextMapper.Update(req, kapi.WithUser(ctx, newUser))
+
+		case userapi.Resource(authorizationapi.UserResource), userapi.Resource(authorizationapi.SystemUserResource):
+			newUser := &user.DefaultInfo{
+				Name: name,
+			}
+			groups, err := c.GroupCache.GroupsFor(name)
+			if err == nil {
+				for _, group := range groups {
+					newUser.Groups = append(newUser.Groups, group.Name)
+				}
+			}
+
+			if resource == userapi.Resource(authorizationapi.UserResource) {
+				newUser.Groups = append(newUser.Groups, "system:authenticated", "system:authenticated:oauth")
```

Comment:
> Are we guaranteed to be an oauth user?

Any user who has a `users` resource *can* have an oauth token.  It seemed more reasonable to put them in this group than not.